src/content/app/tools/blast/utils/blastInputParser.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ParsedInputSequence } from 'src/content/app/tools/blast/types/parsedInputSequence';
+
+/**
+ * Allowed inputs:
+ *
+ * - bare sequence
+ * - bare sequence with numbered lines
+ * - bare sequences separated by empty lines
+ * - FASTA-formatted sequence
+ * - FASTA-formatted multiple sequences (separated by header lines)
+ * - FASTA-formatted sequences separated by empty lines
+ *
+ * Also:
+ * - if a sequence contains spaces, remove them
+ * - make sure all sequence characters are upper-cased
+ */
+
+export const parseBlastInput = (input: string) => {
+  input = input.trim();
+
+  const sequences: ParsedInputSequence[] = [];
+
+  const lines = input.split('\n');
+
+  lines.forEach((line, index) => {
+    line = prepareLine(line);
+
+    if (line.startsWith('>')) {
+      // this is a FASTA header line indicating the beginning of a new sequence
+      line = line.replace(/^\>\s*/, '');
+
+      if (endsInEmptySequenceObject(sequences)) {
+        const sequenceObject = sequences[sequences.length - 1];
+        sequenceObject.header = line; // remove the leading ">" from the header line
+      } else {
+        const newSequence = createEmptySequence();
+        newSequence.header = line;
+        sequences.push(newSequence);
+      }
+    } else if (line === '' && lines[index - 1] !== '') {
+      // interpret it as a sequence separator, and create a new sequence
+      const newSequence = createEmptySequence();
+      sequences.push(newSequence);
+    } else {
+      if (index === 0) {
+        // this is the first line of a bare sequence; create the sequence object first
+        const newSequence = createEmptySequence();
+        sequences.push(newSequence);
+      }
+      const sequenceObject = sequences[sequences.length - 1];
+      sequenceObject.value += line;
+    }
+  });
+
+  return sequences;
+};
+
+const prepareLine = (line: string) => {
+  line = line.trim();
+
+  if (line.startsWith('>')) {
+    // this is a FASTA header; return it as is
+    return line;
+  }
+
+  // remove all digits and spaces from the line
+  line = line.replace(/[\d\s]+/g, '');
+  return line.toUpperCase();
+};
+
+const endsInEmptySequenceObject = (sequences: ParsedInputSequence[]) => {
```

Comment:
How is the name of this function related to what it does?

Reply:
If you refer to the `parseBlastInput` function, you will notice that there is a `sequences` object there that is an array of parsed sequence objects.

As a string passed to the `parseBlastInput` gets parsed line by line:

<img src="https://user-images.githubusercontent.com/6834224/148305484-3b0e8fb0-930e-4a95-a562-4f846b54441c.png" width="400" />

the `sequences` object grows. If the input contains an empty line, the parser will add a new empty sequence object to the sequences array:

<img src="https://user-images.githubusercontent.com/6834224/148305616-86075eb2-2f88-4349-a71b-e2fba1754f54.png" width="600" />

You can see this happening on line 56 of this file.

This is what the `endsInEmptySequenceObject` function is for. It checks if the `sequences` array has an empty sequence object in the end, and acts accordingly — either populates this empty sequence object with data or creates a new empty sequence object and adds data to it.

If there is a better way of keeping track of when/how to add parsed sequences to the array of sequences, I am open to suggestions.

Reply:
I tried the following and it seems to work:
```javascript
const sections = input.split('\n\n');
  sections.forEach((section, index) => {
    const lines = section.split('\n');
    lines.forEach(line => {
      if (line.startsWith('>')) {
        line = line.replace(/^\>\s*/, '');
        sequences[index] = { header: line, value: '' }
      } else if(sequences[index]){
        const sequenceObject = sequences[index];
        sequenceObject.value += line;
      } else {
        sequences[index] = { value: line };
      }
    });
  });

```

Reply:
Sorry, it's this one which passes the tests:

```javascript
  input = input.trim();
  const sequences: ParsedInputSequence[] = [];
  // Replace single line separator with double line 
  input = input.replaceAll('\n\n>', '\n>').replaceAll('\n>', '\n\n>');
  const sections = input.split('\n\n');
  sections.forEach((section, index) => {
    const lines = section.split('\n');
    lines.forEach(line => {
      line = prepareLine(line);
      if (line.startsWith('>')) {
        line = line.replace(/^\>\s*/, '');
        sequences[index] = { header: line, value: '' }
      } else if(sequences[index]){
        const sequenceObject = sequences[index];
        sequenceObject.value += line;
      } else {
        sequences[index] = { value: line };
      }
    });
  });
```

Reply:
I removed the `endsInEmptySequenceObject` checker, but I didn't use `input.replaceAll('\n\n>', '\n>').replaceAll('\n>', '\n\n>');`